sqlglot/serde.py

```diff
@@ -4,65 +4,114 @@
 
 from sqlglot import expressions as exp
 
-if t.TYPE_CHECKING:
-    JSON = t.Union[dict, list, str, float, int, bool, None]
-    Node = t.Union[t.List["Node"], exp.DataType.Type, exp.Expression, JSON]
 
-
-def dump(node: Node) -> JSON:
+def dump(expr: exp.Expression) -> t.Dict[str, t.Any]:
     """
-    Recursively dump an AST into a JSON-serializable dict.
+    Dump an Expression into a JSON serializable dict.
     """
-    if isinstance(node, list):
-        return [dump(i) for i in node]
-    if isinstance(node, exp.DataType.Type):
-        return {
-            "class": "DataType.Type",
-            "value": node.value,
-        }
-    if isinstance(node, exp.Expression):
+
+    root: t.Dict[str, t.Any] = {}
+    stack = [(expr, root)]
+
+    while stack:
+        node, payload = stack.pop()
+
         klass = node.__class__.__qualname__
+
         if node.__class__.__module__ != exp.__name__:
             klass = f"{node.__module__}.{klass}"
-        obj: t.Dict = {
-            "class": klass,
-            "args": {k: dump(v) for k, v in node.args.items() if v is not None and v != []},
-        }
+
+        payload["class"] = klass
+
         if node.type:
-            obj["type"] = dump(node.type)
+            payload["type"] = dump(node.type)
         if node.comments:
-            obj["comments"] = node.comments
+            payload["comments"] = node.comments
         if node._meta is not None:
-            obj["meta"] = node._meta
+            payload["meta"] = node._meta
+        if node.args:
+            args: t.Dict[str, t.Any] = {}
+            payload["args"] = args
+
+            for k, vs in node.args.items():
+                if vs is not None and vs != []:
+                    if hasattr(vs, "parent"):
```

Comment:
Should we do an `isinstance(vs, Expression)` check here instead of relying on a secondary signal?